File: tools/excel/src/main/java/top/bettercode/summer/tools/excel/ExcelErrorHandler.java
package top.bettercode.summer.tools.excel;

import java.time.format.DateTimeParseException;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import javax.servlet.http.HttpServletRequest;
import javax.validation.ConstraintViolation;
import javax.validation.ConstraintViolationException;
import org.springframework.context.MessageSource;
import org.springframework.util.StringUtils;
import top.bettercode.summer.tools.excel.ExcelImportException.CellError;
import top.bettercode.summer.web.RespEntity;
import top.bettercode.summer.web.error.AbstractErrorHandler;

/**
 * @author Peter Wu
 */
public class ExcelErrorHandler extends AbstractErrorHandler {


  public ExcelErrorHandler(MessageSource messageSource,
      HttpServletRequest request) {
    super(messageSource, request);
  }

  @Override
  public void handlerException(Throwable error, RespEntity<?> respEntity,
      Map<String, String> errors, String separator) {
    String message = null;
    if (error instanceof ExcelImportException) {
      List<CellError> cellErrors = ((ExcelImportException) error).getErrors();

      for (CellError cellError : cellErrors) {
        String key = getText(cellError.getMessage(), cellError.getRow(),
            cellError.getColumnName());
        String title = cellError.getTitle();
        Exception value = cellError.getException();
        if (value instanceof ConstraintViolationException) {
          for (ConstraintViolation<?> constraintViolation : ((ConstraintViolationException) value)
              .getConstraintViolations()) {
            errors.put(key, title + ": " + constraintViolation.getMessage());
          }
        } else {
          String msg = value.getMessage();
          if (value instanceof DateTimeParseException) {
            String msgRegex = "Text '(.*?)' could not be parsed at index (\\d+)";
            if (msg.matches(msgRegex)) {
              msg = msg.replaceAll(msgRegex, "$1") + "不是有效的日期格式";
            }
          }
          errors.put(key, title + ": " + getText(msg));
        }
      }
      Entry<String, String> firstError = errors.entrySet().iterator().next();
      message = firstError.getKey() + separator + firstError.getValue();
    }
    if (StringUtils.hasText(message)) {
      respEntity.message = message;
    }
  }
}

File: tools/excel/src/main/kotlin/top/bettercode/summer/tools/excel/ExcelErrorHandler.kt
package top.bettercode.summer.tools.excel

import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.ConstraintViolationException
import org.springframework.context.MessageSource
import top.bettercode.summer.tools.excel.read.ExcelReaderException
import top.bettercode.summer.web.RespEntity
import top.bettercode.summer.web.error.AbstractErrorHandler
import java.time.format.DateTimeParseException
import kotlin.collections.set

/**
 * @author Peter Wu
 */
class ExcelErrorHandler(messageSource: MessageSource,
                        request: HttpServletRequest?) : AbstractErrorHandler(messageSource, request) {
    override fun handlerException(error: Throwable, respEntity: RespEntity<*>,
                                  errors: MutableMap<String?, String?>, separator: String) {
        if (error is ExcelReaderException) {
            val cellErrors = error.errors
            for (cellError in cellErrors) {
                val key = cellError.columnName + cellError.row
                val title = cellError.title
                val exception = cellError.exception
                val value = invalidValue(cellError.value)
                if (exception is ConstraintViolationException) {
                    for (constraintViolation in exception
                            .constraintViolations) {
                        errors[key] = title + ": " + value + constraintViolation.message
                    }
                } else {
                    var msg = exception.message
                    if (exception is DateTimeParseException) {
                        val msgRegex = "Text '(.*?)' could not be parsed at index (\\d+)"
                        if (msg!!.matches(msgRegex.toRegex())) {
                            msg = msg.replace(msgRegex.toRegex(), "$1") + getText("notAValidDateFormat")
                        }
                    }
                    errors[key] = title + ": " + value + getText(msg ?: "unknownError")
                }
            }
            respEntity.message = errors.entries.joinToString {
                it.key + separator + it.value
            }
        }
    }
}
